### PR TITLE
Enable support for BFD on external gateway routes

### PR DESCRIPTION
**Pull request by fedepaol**
**- What this PR does and why is it needed**

When annotating a gateway pod or a namespace with both the external gateways annotation and k8s.ovn.org/bfd-enabled="", the routes to the external gateways are created with the --bfd parameter. This enables the bfd check on those routes.

**- Special notes for reviewers**

This is marked as WIP but ready for review. I want to add e2e tests to cover the feature, but need to rely on the refactoring of external gw tests.

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/egressgw.go [hunk 1]
```diff
@@ -18,13 +18,23 @@ import (
 	"k8s.io/klog/v2"
 )
 
+type gatewayInfo struct {
```
Comment (abhat) [inline go-controller/pkg/ovn/egressgw.go:21]: just curious why we need a separate struct. can't we just pass the boolean around to methods that need bfd based on the annotation?

Comment (fedepaol) [reply go-controller/pkg/ovn/egressgw.go:21]: That was my initial version, but then I realized that the gws are also referenced in the nsInfo structure (see https://github.com/ovn-org/ovn-kubernetes/pull/2077/files#diff-796adb9d8a9b5b5636dc81e7350a00ef143e6e5bcb52ec30dc350c866cd9208cL98 and https://github.com/ovn-org/ovn-kubernetes/pull/2077/files#diff-796adb9d8a9b5b5636dc81e7350a00ef143e6e5bcb52ec30dc350c866cd9208cL106 , where it's contained as map entry). 
Not having a structure would have made the whole thing hard to follow. Each (set of) gws have this property which is bfdEnabled or not, which makes it natural to have it wrapped in its own type.

File: go-controller/pkg/ovn/pods.go [hunk 2]
```diff
@@ -214,16 +214,16 @@ func (oc *Controller) addRoutesGatewayIP(pod *kapi.Pod, podAnnotation *util.PodA
 	return nil
 }
 
-func (oc *Controller) getRoutingExternalGWs(ns string) []net.IP {
+func (oc *Controller) getRoutingExternalGWs(ns string) gatewayInfo {
```
Comment (abhat) [inline go-controller/pkg/ovn/pods.go:217]: this has nothing to do with your code, but we are returning a reference to an object contained within nsInfo here without holding the lock on it?

**Comment by fedepaol** [comment]
E2E tests covered in 
https://github.com/ovn-org/ovn-kubernetes/pull/2091
and 
https://github.com/ovn-org/ovn-kubernetes/pull/2090

File: go-controller/pkg/ovn/pods.go (hunk 2 above)
Comment (fedepaol) [reply go-controller/pkg/ovn/pods.go:217]: Yup maybe it's safer to return a copy?

File: go-controller/pkg/ovn/pods.go [hunk 3]
```diff
@@ -443,17 +443,30 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
...
+	podHasGWRoute := false
 	// if we have any external or pod Gateways, add routes
-	if len(routingExternalGWs) > 0 || len(routingPodGWs) > 0 {
-		routingGWs := routingExternalGWs
-		for _, ipNets := range routingPodGWs {
-			routingGWs = append(routingGWs, ipNets...)
-		}
-		err = oc.addGWRoutesForPod(routingGWs, podIfAddrs, pod.Namespace, pod.Spec.NodeName)
+	if len(routingExternalGWs.gws) > 0 {
+		err = oc.addGWRoutesForPod(routingExternalGWs, podIfAddrs, pod.Namespace, pod.Spec.NodeName)
```
Comment (trozet) [inline go-controller/pkg/ovn/pods.go:449]: is there anyway we can consolidate this and line 461 to a single call for addGWRoutesForPod? We were only calling it once before, and now we are calling it twice. It seems like an expensive function because we have to get a lock and then loop through a bunch of stuff again.

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/pods.go (hunk 2 above)
Comment (fedepaol) [reply go-controller/pkg/ovn/pods.go:217]: In theory, it could be still messed up as the object returned contains a slice, which is "kind of" a reference in go. The array backing up the slice is the same unless some operation happens on one of the copies.
It might be that it's never the case (as it was not until now), but I'd prefer to be on the safe side and copy the slice (in a new result object).

File: go-controller/pkg/ovn/ovn-nbctl --data=bare --no-heading --col [hunk 4]
```diff
@@ -0,0 +1 @@
+ovn-nbctl --data=bare --no-heading --columns=_uuid find BFD dst_ip="9.0.0.2" logical_port=rtoe-GR_ovn-worker
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply go-controller/pkg/ovn/ovn-nbctl --data=bare --no-heading --col:1]: :-)

File: go-controller/pkg/ovn/egressgw.go [hunk 5]
```diff
@@ -438,3 +464,38 @@ func (oc *Controller) delHybridRoutePolicyForPod(podIP net.IP, node string) erro
 	}
 	return nil
 }
+
+func cleanUpBFDEntry(gatewayIP, gatewayRouter string) {
+	portName := types.GWRouterToExtSwitchPrefix + gatewayRouter
+
+	output, stderr, err := util.RunOVNNbctl(
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply go-controller/pkg/ovn/egressgw.go:471]: I thought about that but I wasn't sure it was worth the effort, especially for this case https://github.com/fedepaol/ovn-kubernetes/blob/9ebba002f76bc0df7246dd6ae29b329b96c29509/go-controller/pkg/ovn/egressgw.go#L223 where we don't have the gwInfo object at hand.

This is temporary, when this http://patchwork.ozlabs.org/project/ovn/patch/3c39dc96a36a3445cfa8485a67de79f9f3d5651b.1614602770.git.lorenzo.bianconi@redhat.com/ gets in, we won't need it anymore.

I'll add the doctext, but I'd avoid the cache thing because it adds extra (un-needed, I think) complexity

File: go-controller/pkg/ovn/egressgw.go [hunk 6]
```diff
@@ -438,3 +464,38 @@ func (oc *Controller) delHybridRoutePolicyForPod(podIP net.IP, node string) erro
...
+		klog.Errorf("cleanUpBFDEntry: failed to list routes for %s, stderr: %q, (%v)", portName, gatewayIP, err, stderr)
+		return
+	}
+	// the bfd entry is still referenced, meaning there's another route on the router
+	// referencing it.
+	if strings.TrimSpace(output) != "" {
+		return
+	}
+	uuid, stderr, err := util.RunOVNNbctl(
+		"--format=csv", "--data=bare", "--no-heading", "--columns=_uuid", "find", "BFD", "logical_port="+portName, "dst_ip="+gatewayIP)
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply go-controller/pkg/ovn/egressgw.go:484]: ok

File: go-controller/pkg/ovn/pods.go (hunk 3 above)
Comment (fedepaol) [reply go-controller/pkg/ovn/pods.go:449]: Yup will change it somehow to accept the two gws. 
I missed the loop over podIfAddrs, makes sense to have it only once.

File: go-controller/pkg/ovn/egressgw.go [hunk 7]
```diff
@@ -438,3 +464,38 @@ func (oc *Controller) delHybridRoutePolicyForPod(podIP net.IP, node string) erro
 	}
 	return nil
 }
+
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply go-controller/pkg/ovn/egressgw.go:467]: Ok

File: go-controller/pkg/ovn/egressgw.go [hunk 8]
```diff
@@ -74,44 +84,52 @@ func (oc *Controller) addPodExternalGW(pod *kapi.Pod) error {
...
+		for _, gw := range egress.gws {
 			for _, podIP := range pod.Status.PodIPs {
 				if utilnet.IsIPv6(gw) != utilnet.IsIPv6String(podIP.IP) {
 					continue
 				}
+
 				mask := GetIPFullMask(podIP.IP)
-				_, stderr, err := util.RunOVNNbctl("--", "--may-exist", "--policy=src-ip", "--ecmp-symmetric-reply",
-					"lr-route-add", gr, podIP.IP+mask, gw.String())
+				nbctlArgs := []string{"--may-exist", "--policy=src-ip", "--ecmp-symmetric-reply",
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply go-controller/pkg/ovn/egressgw.go:124]: The bfd variant also wants the portname (otherwise it returns an error) while the other one does not need it.

File: go-controller/pkg/ovn/pods.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/ovn/pods.go:217]: yeah thats a good point

**Review by trozet — APPROVED** [review]
/lgtm
